extmod/modussl_mbedtls.c: Add missing args and constants to ssl module.

Carglglz opened the pull request:
This PR adds `cert_reqs` , `ca_certs` arguments to `ssl.wrap_socket` and 
`ssl.CERT_NONE`, `ssl.CERT_OPTIONAL`, `ssl.CERT_REQUIRED` constants to ssl module following documentation.

This enables mutual authentication (mTLS), so it possible to authenticate both servers and clients.

Also I'm aware there is a previous draft of this at #5998 however there the authentication mode is hard coded to optional (`MBEDTLS_SSL_VERIFY_OPTIONAL`) which may not be [recommended](https://tls.mbed.org/api/ssl_8h.html#a5695285c9dbfefec295012b566290f37) 

>On client, MBEDTLS_SSL_VERIFY_REQUIRED is the recommended mode. With   MBEDTLS_SSL_VERIFY_OPTIONAL, the user needs to call mbedtls_ssl_get_verify_result() at the right time(s), which may not be obvious, while REQUIRED always perform the verification as soon as possible. For example, REQUIRED was protecting against the "triple handshake" attack even before it was found.

I'm not sure if these lines could be just one line or if it is better this way?
```
int certreq = (args->cert_reqs.u_int);
mbedtls_ssl_conf_authmode(&o->conf, certreq);
```
Also here 
```
{ MP_QSTR_cert_reqs, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0}}
````
It may be better use `MBEDTLS_SSL_VERIFY_NONE` for clarity?

Carglglz committed:
extmod/modussl_mbedtls.c: Add missing args and constants to ssl module.

Add cert_reqs and ca_certs kwargs to ssl.wrap_socket and
ssl.CERT_NONE, ssl.CERT_OPTIONAL, ssl.CERT_REQUIRED options
to allow certificate validation.

Signed-off-by: Carlos Gil <carlosgilglez@gmail.com>

dpgeorge commented:
> It may be better use `MBEDTLS_SSL_VERIFY_NONE` for clarity?

Yes I think that would be good.  Is that the same default that CPython has?

As mentioned in #8854, I'd prefer to merge this PR for the core `ca_certs` support.  Are you able to add docs and tests for this new feature?  Document the 2 new keyword args, and add a test similar to that in #5998.

Carglglz commented:
> Are you able to add docs and tests for this new feature? Document the 2 new keyword args, and add a test similar to that in https://github.com/micropython/micropython/pull/5998.

Docs I can for sure, tests I'm working on it.
So far self-signed key, cert works but certs from `google.com` or `micropython.org` fails with 
```
Address infos: [(2, 1, 0, 'micropython.org', ('176.58.119.26', 443))]
Connect address: ('176.58.119.26', 443)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "http_ssl_test.py", line 60, in <module>
  File "http_ssl_test.py", line 49, in main
OSError: (-9984, 'MBEDTLS_ERR_X509_CERT_VERIFY_FAILED')
```

For self-signed key, cert I tested ECC and RSA:

## ECC key, cert

```
$ openssl ecparam -out ec-cakey.pem -name prime256v1 -genkey
$ openssl req -new -x509 -days 3650 -key ec-cakey.pem -out ec-cacert.pem
```
Then I used a modified version of `test/multi_net/ssl_data.py` 

`ssl_ecc_auth.py`
```
# Simple test creating an SSL connection and transferring some data
# This test won't run under CPython because it requires key/cert

try:
    import ubinascii as binascii, usocket as socket, ussl as ssl
except ImportError:
    print("SKIP")
    raise SystemExit

PORT = 8000

# This self-signed key/cert pair is randomly generated and to be used for
# testing/demonstration only.  You should always generate your own key/cert.
cert = b"""-----BEGIN CERTIFICATE-----
MIICSTCCAe+gAwIBAgIJAK7kDH3KW/pbMAoGCCqGSM49BAMCMIGAMQswCQYDVQQG
EwJBVTEMMAoGA1UECAwDRm9vMQwwCgYDVQQHDANCYXIxFDASBgNVBAoMC01pY3Jv
UHl0aG9uMQswCQYDVQQLDAJNUDEWMBQGA1UEAwwNZXNwaG9tZS5sb2NhbDEaMBgG
CSqGSIb3DQEJARYLZm9vQGJhci5jb20wHhcNMjIwNzEyMTcxOTU5WhcNMzIwNzA5
MTcxOTU5WjCBgDELMAkGA1UEBhMCQVUxDDAKBgNVBAgMA0ZvbzEMMAoGA1UEBwwD
QmFyMRQwEgYDVQQKDAtNaWNyb1B5dGhvbjELMAkGA1UECwwCTVAxFjAUBgNVBAMM
DWVzcGhvbWUubG9jYWwxGjAYBgkqhkiG9w0BCQEWC2Zvb0BiYXIuY29tMFkwEwYH
KoZIzj0CAQYIKoZIzj0DAQcDQgAEmuTTWsQ44oUgEt/fpjm33x2JbS8FdiVnC2Cc
4QFUsr8AqVTOwCzqqDO6MvXask8WyBjmROV0NS1QAp8zM8CnlqNQME4wHQYDVR0O
BBYEFLS6piYUhW4Z3l6DSyNC6BuJetg9MB8GA1UdIwQYMBaAFLS6piYUhW4Z3l6D
SyNC6BuJetg9MAwGA1UdEwQFMAMBAf8wCgYIKoZIzj0EAwIDSAAwRQIgaoRjizC+
AvvlQQjl6Jw3lFCxvuYrwy4YwNKgqJDyz8UCIQCFyMWi4kkniUZxhCZwuvJiPXl/
HFSxkQyQfayCKkApCA==
-----END CERTIFICATE-----"""

key = b"""-----BEGIN EC PARAMETERS-----
BggqhkjOPQMBBw==
-----END EC PARAMETERS-----
-----BEGIN EC PRIVATE KEY-----
MHcCAQEEII9I4GhoeGqhgOz5Ybqh5NTEwLQmli1jIzvrr6/6clLSoAoGCCqGSM49
AwEHoUQDQgAEmuTTWsQ44oUgEt/fpjm33x2JbS8FdiVnC2Cc4QFUsr8AqVTOwCzq
qDO6MvXask8WyBjmROV0NS1QAp8zM8Cnlg==
-----END EC PRIVATE KEY-----"""


# Server
def instance0():

    s = socket.socket()
    s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    s.bind(socket.getaddrinfo("0.0.0.0", PORT)[0][-1])
    s.listen(1)

    s2, _ = s.accept()
    s2 = ssl.wrap_socket(s2, server_side=True, key=key, cert=cert)
    print(s2.read(16))
    s2.write(b"server to client")
    s2.close()
    s.close()


# Client
def instance1():

    s = socket.socket()
    s.connect(socket.getaddrinfo('esphome.local', PORT)[0][-1])
    s = ssl.wrap_socket(s, cert_reqs=ssl.CERT_REQUIRED, ca_certs=cert)
    s.write(b"client to server")
    print(s.read(16))
    s.close()

```

One of the esp32 is running `instance0()`  (server with hostname `esphome.local` )
and the other  `instance1()`

### Server
```
>>> from ssl_ecc_auth import instance0
>>> instance0()
b'client to server'
>>>
```
### Client

```
>>> from ssl_ecc_auth import instance1
>>> instance1()
b'server to client'
>>>
```

This works too with RSA

### RSA key, cert
```
$ openssl req -x509 -newkey rsa:4096 -keyout rsa_key.pem -out rsa_cert.pem -days 365 -node
````
The rest is the same I've just changed key, cert and named the test `ssl_rsa_auth.py`

Carglglz commented:
Ok I figured out what I was doing wrong, I wasn't using the entire cert chain or at least the correct cert.

This works for `micropython.org`

I get the cert chain with
 ```
$ openssl s_client -showcerts -connect micropython.org:443 </dev/null 2>/dev/null
```

```
CONNECTED(00000003)
---
Certificate chain
 0 s:/CN=dpgeorge.net
   i:/C=US/O=Let's Encrypt/CN=R3
-----BEGIN CERTIFICATE-----
MIIFHzCCBAegAwIBAgISBDNT3d/NtE1ppCrdRpyVve9rMA0GCSqGSIb3DQEBCwUA
MDIxCzAJBgNVBAYTAlVTMRYwFAYDVQQKEw1MZXQncyBFbmNyeXB0MQswCQYDVQQD
EwJSMzAeFw0yMjA1MjYwNzQ1MTJaFw0yMjA4MjQwNzQ1MTFaMBcxFTATBgNVBAMT
DGRwZ2VvcmdlLm5ldDCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALt3
Kx60TijbUvhOSCgCY4BDwYcmTTKoBm9bpXrhaYoCJEX05lCXghnEuWvudjvRsfuK
paKo1i03lzCcdOoAorcqlwTVO09XKdEAtNi7bBpbr8ajqB7ptRm9aGLWi2p0lTGp
+uyNY6LUjNgCf/VJvFddPOS8q0TNhakdgG8BR31PDAxuWy3wA2YTv8Ppj089elEY
UmPhFIm1YSnMH3Owjtukwv0r3pe03AJ760Lgm4hHZO2S0H/qH0gyoWkHbq+e7Ba8
4gtursn7M6E2vWBNgFsV1fZTyIRmyWBRgIVSIxGU4ultzmfJQI/hRgD8cnOnZrrC
Ns3+1bwFMXSiuKIhBU0CAwEAAaOCAkgwggJEMA4GA1UdDwEB/wQEAwIFoDAdBgNV
HSUEFjAUBggrBgEFBQcDAQYIKwYBBQUHAwIwDAYDVR0TAQH/BAIwADAdBgNVHQ4E
FgQUHlWp/zVL6yP4hjPIT5MgkdnuU6wwHwYDVR0jBBgwFoAUFC6zF7dYVsuuUAlA
5h+vnYsUwsYwVQYIKwYBBQUHAQEESTBHMCEGCCsGAQUFBzABhhVodHRwOi8vcjMu
by5sZW5jci5vcmcwIgYIKwYBBQUHMAKGFmh0dHA6Ly9yMy5pLmxlbmNyLm9yZy8w
FwYDVR0RBBAwDoIMZHBnZW9yZ2UubmV0MEwGA1UdIARFMEMwCAYGZ4EMAQIBMDcG
CysGAQQBgt8TAQEBMCgwJgYIKwYBBQUHAgEWGmh0dHA6Ly9jcHMubGV0c2VuY3J5
cHQub3JnMIIBBQYKKwYBBAHWeQIEAgSB9gSB8wDxAHcAKXm+8J45OSHwVnOfY6V3
5b5XfZxgCvj5TV0mXCVdx4QAAAGA/4qQPwAABAMASDBGAiEAlvSRyAAttg6WPIdn
hKh8ErL9qJBDq3OtcHntRxMCTAsCIQDhofbSXGUEyIe9pdavm+lMQUn3Pm5NjApj
sZC0klwBpgB2AG9Tdqwx8DEZ2JkApFEV/3cVHBHZAsEAKQaNsgiaN9kTAAABgP+K
kb0AAAQDAEcwRQIhAK29v0DD0yJ+m9pTfmRl+7f81NBJhAoEOUIwP7reiCxFAiBU
+JL/eDTynJYoP4ApO0Fkld6K7NpssZQf8ORlwVs/STANBgkqhkiG9w0BAQsFAAOC
AQEAVBMxcjf0FdQrsLetAZBkWi6ONVQYzTA1YSOzP1WMuenmd6uQY+j50iJv7RPX
KW4tmI0qjIWU32xlFxZLrZyWiplUXt4Vq19YvEygWk9pzjR4QBr/lmsk5YPKfY9u
z7faFDn/r4UrUdgGtaqO4HpVoXSQ9TGs9hYwdXyJPfIvsX431g8K+Fw/Kd9g3vlQ
IVFcjsVN9dafdz3e5VvKCXrZOqct5WbhcYqkfm4ehl7/PA16D0Y8b/8uWzemWxut
T1laVGkPtQuPqd1ZdralR1Jsjk6CgxhGuEH60u4xY7QwyRcS83951z/6jJPIE5Ss
H3+JgRraIhzr2urkEQaedBPK/g==
-----END CERTIFICATE-----
 1 s:/C=US/O=Let's Encrypt/CN=R3
   i:/C=US/O=Internet Security Research Group/CN=ISRG Root X1
-----BEGIN CERTIFICATE-----
MIIFFjCCAv6gAwIBAgIRAJErCErPDBinU/bWLiWnX1owDQYJKoZIhvcNAQELBQAw
TzELMAkGA1UEBhMCVVMxKTAnBgNVBAoTIEludGVybmV0IFNlY3VyaXR5IFJlc2Vh
cmNoIEdyb3VwMRUwEwYDVQQDEwxJU1JHIFJvb3QgWDEwHhcNMjAwOTA0MDAwMDAw
WhcNMjUwOTE1MTYwMDAwWjAyMQswCQYDVQQGEwJVUzEWMBQGA1UEChMNTGV0J3Mg
RW5jcnlwdDELMAkGA1UEAxMCUjMwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEK
AoIBAQC7AhUozPaglNMPEuyNVZLD+ILxmaZ6QoinXSaqtSu5xUyxr45r+XXIo9cP
R5QUVTVXjJ6oojkZ9YI8QqlObvU7wy7bjcCwXPNZOOftz2nwWgsbvsCUJCWH+jdx
sxPnHKzhm+/b5DtFUkWWqcFTzjTIUu61ru2P3mBw4qVUq7ZtDpelQDRrK9O8Zutm
NHz6a4uPVymZ+DAXXbpyb/uBxa3Shlg9F8fnCbvxK/eG3MHacV3URuPMrSXBiLxg
Z3Vms/EY96Jc5lP/Ooi2R6X/ExjqmAl3P51T+c8B5fWmcBcUr2Ok/5mzk53cU6cG
/kiFHaFpriV1uxPMUgP17VGhi9sVAgMBAAGjggEIMIIBBDAOBgNVHQ8BAf8EBAMC
AYYwHQYDVR0lBBYwFAYIKwYBBQUHAwIGCCsGAQUFBwMBMBIGA1UdEwEB/wQIMAYB
Af8CAQAwHQYDVR0OBBYEFBQusxe3WFbLrlAJQOYfr52LFMLGMB8GA1UdIwQYMBaA
FHm0WeZ7tuXkAXOACIjIGlj26ZtuMDIGCCsGAQUFBwEBBCYwJDAiBggrBgEFBQcw
AoYWaHR0cDovL3gxLmkubGVuY3Iub3JnLzAnBgNVHR8EIDAeMBygGqAYhhZodHRw
Oi8veDEuYy5sZW5jci5vcmcvMCIGA1UdIAQbMBkwCAYGZ4EMAQIBMA0GCysGAQQB
gt8TAQEBMA0GCSqGSIb3DQEBCwUAA4ICAQCFyk5HPqP3hUSFvNVneLKYY611TR6W
PTNlclQtgaDqw+34IL9fzLdwALduO/ZelN7kIJ+m74uyA+eitRY8kc607TkC53wl
ikfmZW4/RvTZ8M6UK+5UzhK8jCdLuMGYL6KvzXGRSgi3yLgjewQtCPkIVz6D2QQz
CkcheAmCJ8MqyJu5zlzyZMjAvnnAT45tRAxekrsu94sQ4egdRCnbWSDtY7kh+BIm
lJNXoB1lBMEKIq4QDUOXoRgffuDghje1WrG9ML+Hbisq/yFOGwXD9RiX8F6sw6W4
avAuvDszue5L3sz85K+EC4Y/wFVDNvZo4TYXao6Z0f+lQKc0t8DQYzk1OXVu8rp2
yJMC6alLbBfODALZvYH7n7do1AZls4I9d1P4jnkDrQoxB3UqQ9hVl3LEKQ73xF1O
yK5GhDDX8oVfGKF5u+decIsH4YaTw7mP3GFxJSqv3+0lUFJoi5Lc5da149p90Ids
hCExroL1+7mryIkXPeFM5TgO9r0rvZaBFOvV2z0gp35Z0+L4WPlbuEjN/lxPFin+
HlUjr8gRsI3qfJOQFy/9rKIJR0Y/8Omwt/8oTWgy1mdeHmmjk7j1nYsvC9JSQ6Zv
MldlTTKB3zhThV1+XWYp6rjd5JW1zbVWEkLNxE7GJThEUG3szgBVGP7pSWTUTsqX
nLRbwHOoq7hHwg==
-----END CERTIFICATE-----
 2 s:/C=US/O=Internet Security Research Group/CN=ISRG Root X1
   i:/O=Digital Signature Trust Co./CN=DST Root CA X3
-----BEGIN CERTIFICATE-----
MIIFYDCCBEigAwIBAgIQQAF3ITfU6UK47naqPGQKtzANBgkqhkiG9w0BAQsFADA/
MSQwIgYDVQQKExtEaWdpdGFsIFNpZ25hdHVyZSBUcnVzdCBDby4xFzAVBgNVBAMT
DkRTVCBSb290IENBIFgzMB4XDTIxMDEyMDE5MTQwM1oXDTI0MDkzMDE4MTQwM1ow
TzELMAkGA1UEBhMCVVMxKTAnBgNVBAoTIEludGVybmV0IFNlY3VyaXR5IFJlc2Vh
cmNoIEdyb3VwMRUwEwYDVQQDEwxJU1JHIFJvb3QgWDEwggIiMA0GCSqGSIb3DQEB
AQUAA4ICDwAwggIKAoICAQCt6CRz9BQ385ueK1coHIe+3LffOJCMbjzmV6B493XC
ov71am72AE8o295ohmxEk7axY/0UEmu/H9LqMZshftEzPLpI9d1537O4/xLxIZpL
wYqGcWlKZmZsj348cL+tKSIG8+TA5oCu4kuPt5l+lAOf00eXfJlII1PoOK5PCm+D
LtFJV4yAdLbaL9A4jXsDcCEbdfIwPPqPrt3aY6vrFk/CjhFLfs8L6P+1dy70sntK
4EwSJQxwjQMpoOFTJOwT2e4ZvxCzSow/iaNhUd6shweU9GNx7C7ib1uYgeGJXDR5
bHbvO5BieebbpJovJsXQEOEO3tkQjhb7t/eo98flAgeYjzYIlefiN5YNNnWe+w5y
sR2bvAP5SQXYgd0FtCrWQemsAXaVCg/Y39W9Eh81LygXbNKYwagJZHduRze6zqxZ
Xmidf3LWicUGQSk+WT7dJvUkyRGnWqNMQB9GoZm1pzpRboY7nn1ypxIFeFntPlF4
FQsDj43QLwWyPntKHEtzBRL8xurgUBN8Q5N0s8p0544fAQjQMNRbcTa0B7rBMDBc
SLeCO5imfWCKoqMpgsy6vYMEG6KDA0Gh1gXxG8K28Kh8hjtGqEgqiNx2mna/H2ql
PRmP6zjzZN7IKw0KKP/32+IVQtQi0Cdd4Xn+GOdwiK1O5tmLOsbdJ1Fu/7xk9TND
TwIDAQABo4IBRjCCAUIwDwYDVR0TAQH/BAUwAwEB/zAOBgNVHQ8BAf8EBAMCAQYw
SwYIKwYBBQUHAQEEPzA9MDsGCCsGAQUFBzAChi9odHRwOi8vYXBwcy5pZGVudHJ1
c3QuY29tL3Jvb3RzL2RzdHJvb3RjYXgzLnA3YzAfBgNVHSMEGDAWgBTEp7Gkeyxx
+tvhS5B1/8QVYIWJEDBUBgNVHSAETTBLMAgGBmeBDAECATA/BgsrBgEEAYLfEwEB
ATAwMC4GCCsGAQUFBwIBFiJodHRwOi8vY3BzLnJvb3QteDEubGV0c2VuY3J5cHQu
b3JnMDwGA1UdHwQ1MDMwMaAvoC2GK2h0dHA6Ly9jcmwuaWRlbnRydXN0LmNvbS9E
U1RST09UQ0FYM0NSTC5jcmwwHQYDVR0OBBYEFHm0WeZ7tuXkAXOACIjIGlj26Ztu
MA0GCSqGSIb3DQEBCwUAA4IBAQAKcwBslm7/DlLQrt2M51oGrS+o44+/yQoDFVDC
5WxCu2+b9LRPwkSICHXM6webFGJueN7sJ7o5XPWioW5WlHAQU7G75K/QosMrAdSW
9MUgNTP52GE24HGNtLi1qoJFlcDyqSMo59ahy2cI2qBDLKobkx/J3vWraV0T9VuG
WCLKTVXkcGdtwlfFRjlBz4pYg1htmf5X6DYO8A4jqv2Il9DjXA6USbW1FzXSLr9O
he8Y4IWS6wY7bCkjCWDcRQJMEhg76fsO3txE+FiYruq9RUWhiF1myv4Q6W+CyBFC
Dfvp7OOGAN6dEOM4+qR9sdjoSYKEBpsr6GtPAQw4dy753ec5
-----END CERTIFICATE-----
---
Server certificate
subject=/CN=dpgeorge.net
issuer=/C=US/O=Let's Encrypt/CN=R3
---
No client certificate CA names sent
Peer signing digest: SHA512
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 4514 bytes and written 434 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: DD045EDAAF7B29104259EE8B6B3203EFA8860A4EA69127FC8752A59CA8F20147
    Session-ID-ctx:
    Master-Key: EC8F93A4BC345A8695D194DDBF066BC66B1A32A844DA775390B8F774B923A3F3AF65C53A2A17DE957BAD4F9CDCDA0614
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1657655555
    Timeout   : 300 (sec)
    Verify return code: 20 (unable to get local issuer certificate)
---
```

Now `http_ssl_auth_test.py` works.

```
try:
    import usocket as _socket
except:
    import _socket
try:
    import ussl as ssl
except:
    import ssl


ca_cert_chain = b"""-----BEGIN CERTIFICATE-----
MIIFHzCCBAegAwIBAgISBDNT3d/NtE1ppCrdRpyVve9rMA0GCSqGSIb3DQEBCwUA
MDIxCzAJBgNVBAYTAlVTMRYwFAYDVQQKEw1MZXQncyBFbmNyeXB0MQswCQYDVQQD
EwJSMzAeFw0yMjA1MjYwNzQ1MTJaFw0yMjA4MjQwNzQ1MTFaMBcxFTATBgNVBAMT
DGRwZ2VvcmdlLm5ldDCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALt3
Kx60TijbUvhOSCgCY4BDwYcmTTKoBm9bpXrhaYoCJEX05lCXghnEuWvudjvRsfuK
paKo1i03lzCcdOoAorcqlwTVO09XKdEAtNi7bBpbr8ajqB7ptRm9aGLWi2p0lTGp
+uyNY6LUjNgCf/VJvFddPOS8q0TNhakdgG8BR31PDAxuWy3wA2YTv8Ppj089elEY
UmPhFIm1YSnMH3Owjtukwv0r3pe03AJ760Lgm4hHZO2S0H/qH0gyoWkHbq+e7Ba8
4gtursn7M6E2vWBNgFsV1fZTyIRmyWBRgIVSIxGU4ultzmfJQI/hRgD8cnOnZrrC
Ns3+1bwFMXSiuKIhBU0CAwEAAaOCAkgwggJEMA4GA1UdDwEB/wQEAwIFoDAdBgNV
HSUEFjAUBggrBgEFBQcDAQYIKwYBBQUHAwIwDAYDVR0TAQH/BAIwADAdBgNVHQ4E
FgQUHlWp/zVL6yP4hjPIT5MgkdnuU6wwHwYDVR0jBBgwFoAUFC6zF7dYVsuuUAlA
5h+vnYsUwsYwVQYIKwYBBQUHAQEESTBHMCEGCCsGAQUFBzABhhVodHRwOi8vcjMu
by5sZW5jci5vcmcwIgYIKwYBBQUHMAKGFmh0dHA6Ly9yMy5pLmxlbmNyLm9yZy8w
FwYDVR0RBBAwDoIMZHBnZW9yZ2UubmV0MEwGA1UdIARFMEMwCAYGZ4EMAQIBMDcG
CysGAQQBgt8TAQEBMCgwJgYIKwYBBQUHAgEWGmh0dHA6Ly9jcHMubGV0c2VuY3J5
cHQub3JnMIIBBQYKKwYBBAHWeQIEAgSB9gSB8wDxAHcAKXm+8J45OSHwVnOfY6V3
5b5XfZxgCvj5TV0mXCVdx4QAAAGA/4qQPwAABAMASDBGAiEAlvSRyAAttg6WPIdn
hKh8ErL9qJBDq3OtcHntRxMCTAsCIQDhofbSXGUEyIe9pdavm+lMQUn3Pm5NjApj
sZC0klwBpgB2AG9Tdqwx8DEZ2JkApFEV/3cVHBHZAsEAKQaNsgiaN9kTAAABgP+K
kb0AAAQDAEcwRQIhAK29v0DD0yJ+m9pTfmRl+7f81NBJhAoEOUIwP7reiCxFAiBU
+JL/eDTynJYoP4ApO0Fkld6K7NpssZQf8ORlwVs/STANBgkqhkiG9w0BAQsFAAOC
AQEAVBMxcjf0FdQrsLetAZBkWi6ONVQYzTA1YSOzP1WMuenmd6uQY+j50iJv7RPX
KW4tmI0qjIWU32xlFxZLrZyWiplUXt4Vq19YvEygWk9pzjR4QBr/lmsk5YPKfY9u
z7faFDn/r4UrUdgGtaqO4HpVoXSQ9TGs9hYwdXyJPfIvsX431g8K+Fw/Kd9g3vlQ
IVFcjsVN9dafdz3e5VvKCXrZOqct5WbhcYqkfm4ehl7/PA16D0Y8b/8uWzemWxut
T1laVGkPtQuPqd1ZdralR1Jsjk6CgxhGuEH60u4xY7QwyRcS83951z/6jJPIE5Ss
H3+JgRraIhzr2urkEQaedBPK/g==
-----END CERTIFICATE-----
 1 s:/C=US/O=Let's Encrypt/CN=R3
   i:/C=US/O=Internet Security Research Group/CN=ISRG Root X1
-----BEGIN CERTIFICATE-----
MIIFFjCCAv6gAwIBAgIRAJErCErPDBinU/bWLiWnX1owDQYJKoZIhvcNAQELBQAw
TzELMAkGA1UEBhMCVVMxKTAnBgNVBAoTIEludGVybmV0IFNlY3VyaXR5IFJlc2Vh
cmNoIEdyb3VwMRUwEwYDVQQDEwxJU1JHIFJvb3QgWDEwHhcNMjAwOTA0MDAwMDAw
WhcNMjUwOTE1MTYwMDAwWjAyMQswCQYDVQQGEwJVUzEWMBQGA1UEChMNTGV0J3Mg
RW5jcnlwdDELMAkGA1UEAxMCUjMwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEK
AoIBAQC7AhUozPaglNMPEuyNVZLD+ILxmaZ6QoinXSaqtSu5xUyxr45r+XXIo9cP
R5QUVTVXjJ6oojkZ9YI8QqlObvU7wy7bjcCwXPNZOOftz2nwWgsbvsCUJCWH+jdx
sxPnHKzhm+/b5DtFUkWWqcFTzjTIUu61ru2P3mBw4qVUq7ZtDpelQDRrK9O8Zutm
NHz6a4uPVymZ+DAXXbpyb/uBxa3Shlg9F8fnCbvxK/eG3MHacV3URuPMrSXBiLxg
Z3Vms/EY96Jc5lP/Ooi2R6X/ExjqmAl3P51T+c8B5fWmcBcUr2Ok/5mzk53cU6cG
/kiFHaFpriV1uxPMUgP17VGhi9sVAgMBAAGjggEIMIIBBDAOBgNVHQ8BAf8EBAMC
AYYwHQYDVR0lBBYwFAYIKwYBBQUHAwIGCCsGAQUFBwMBMBIGA1UdEwEB/wQIMAYB
Af8CAQAwHQYDVR0OBBYEFBQusxe3WFbLrlAJQOYfr52LFMLGMB8GA1UdIwQYMBaA
FHm0WeZ7tuXkAXOACIjIGlj26ZtuMDIGCCsGAQUFBwEBBCYwJDAiBggrBgEFBQcw
AoYWaHR0cDovL3gxLmkubGVuY3Iub3JnLzAnBgNVHR8EIDAeMBygGqAYhhZodHRw
Oi8veDEuYy5sZW5jci5vcmcvMCIGA1UdIAQbMBkwCAYGZ4EMAQIBMA0GCysGAQQB
gt8TAQEBMA0GCSqGSIb3DQEBCwUAA4ICAQCFyk5HPqP3hUSFvNVneLKYY611TR6W
PTNlclQtgaDqw+34IL9fzLdwALduO/ZelN7kIJ+m74uyA+eitRY8kc607TkC53wl
ikfmZW4/RvTZ8M6UK+5UzhK8jCdLuMGYL6KvzXGRSgi3yLgjewQtCPkIVz6D2QQz
CkcheAmCJ8MqyJu5zlzyZMjAvnnAT45tRAxekrsu94sQ4egdRCnbWSDtY7kh+BIm
lJNXoB1lBMEKIq4QDUOXoRgffuDghje1WrG9ML+Hbisq/yFOGwXD9RiX8F6sw6W4
avAuvDszue5L3sz85K+EC4Y/wFVDNvZo4TYXao6Z0f+lQKc0t8DQYzk1OXVu8rp2
yJMC6alLbBfODALZvYH7n7do1AZls4I9d1P4jnkDrQoxB3UqQ9hVl3LEKQ73xF1O
yK5GhDDX8oVfGKF5u+decIsH4YaTw7mP3GFxJSqv3+0lUFJoi5Lc5da149p90Ids
hCExroL1+7mryIkXPeFM5TgO9r0rvZaBFOvV2z0gp35Z0+L4WPlbuEjN/lxPFin+
HlUjr8gRsI3qfJOQFy/9rKIJR0Y/8Omwt/8oTWgy1mdeHmmjk7j1nYsvC9JSQ6Zv
MldlTTKB3zhThV1+XWYp6rjd5JW1zbVWEkLNxE7GJThEUG3szgBVGP7pSWTUTsqX
nLRbwHOoq7hHwg==
-----END CERTIFICATE-----
 2 s:/C=US/O=Internet Security Research Group/CN=ISRG Root X1
   i:/O=Digital Signature Trust Co./CN=DST Root CA X3
-----BEGIN CERTIFICATE-----
MIIFYDCCBEigAwIBAgIQQAF3ITfU6UK47naqPGQKtzANBgkqhkiG9w0BAQsFADA/
MSQwIgYDVQQKExtEaWdpdGFsIFNpZ25hdHVyZSBUcnVzdCBDby4xFzAVBgNVBAMT
DkRTVCBSb290IENBIFgzMB4XDTIxMDEyMDE5MTQwM1oXDTI0MDkzMDE4MTQwM1ow
TzELMAkGA1UEBhMCVVMxKTAnBgNVBAoTIEludGVybmV0IFNlY3VyaXR5IFJlc2Vh
cmNoIEdyb3VwMRUwEwYDVQQDEwxJU1JHIFJvb3QgWDEwggIiMA0GCSqGSIb3DQEB
AQUAA4ICDwAwggIKAoICAQCt6CRz9BQ385ueK1coHIe+3LffOJCMbjzmV6B493XC
ov71am72AE8o295ohmxEk7axY/0UEmu/H9LqMZshftEzPLpI9d1537O4/xLxIZpL
wYqGcWlKZmZsj348cL+tKSIG8+TA5oCu4kuPt5l+lAOf00eXfJlII1PoOK5PCm+D
LtFJV4yAdLbaL9A4jXsDcCEbdfIwPPqPrt3aY6vrFk/CjhFLfs8L6P+1dy70sntK
4EwSJQxwjQMpoOFTJOwT2e4ZvxCzSow/iaNhUd6shweU9GNx7C7ib1uYgeGJXDR5
bHbvO5BieebbpJovJsXQEOEO3tkQjhb7t/eo98flAgeYjzYIlefiN5YNNnWe+w5y
sR2bvAP5SQXYgd0FtCrWQemsAXaVCg/Y39W9Eh81LygXbNKYwagJZHduRze6zqxZ
Xmidf3LWicUGQSk+WT7dJvUkyRGnWqNMQB9GoZm1pzpRboY7nn1ypxIFeFntPlF4
FQsDj43QLwWyPntKHEtzBRL8xurgUBN8Q5N0s8p0544fAQjQMNRbcTa0B7rBMDBc
SLeCO5imfWCKoqMpgsy6vYMEG6KDA0Gh1gXxG8K28Kh8hjtGqEgqiNx2mna/H2ql
PRmP6zjzZN7IKw0KKP/32+IVQtQi0Cdd4Xn+GOdwiK1O5tmLOsbdJ1Fu/7xk9TND
TwIDAQABo4IBRjCCAUIwDwYDVR0TAQH/BAUwAwEB/zAOBgNVHQ8BAf8EBAMCAQYw
SwYIKwYBBQUHAQEEPzA9MDsGCCsGAQUFBzAChi9odHRwOi8vYXBwcy5pZGVudHJ1
c3QuY29tL3Jvb3RzL2RzdHJvb3RjYXgzLnA3YzAfBgNVHSMEGDAWgBTEp7Gkeyxx
+tvhS5B1/8QVYIWJEDBUBgNVHSAETTBLMAgGBmeBDAECATA/BgsrBgEEAYLfEwEB
ATAwMC4GCCsGAQUFBwIBFiJodHRwOi8vY3BzLnJvb3QteDEubGV0c2VuY3J5cHQu
b3JnMDwGA1UdHwQ1MDMwMaAvoC2GK2h0dHA6Ly9jcmwuaWRlbnRydXN0LmNvbS9E
U1RST09UQ0FYM0NSTC5jcmwwHQYDVR0OBBYEFHm0WeZ7tuXkAXOACIjIGlj26Ztu
MA0GCSqGSIb3DQEBCwUAA4IBAQAKcwBslm7/DlLQrt2M51oGrS+o44+/yQoDFVDC
5WxCu2+b9LRPwkSICHXM6webFGJueN7sJ7o5XPWioW5WlHAQU7G75K/QosMrAdSW
9MUgNTP52GE24HGNtLi1qoJFlcDyqSMo59ahy2cI2qBDLKobkx/J3vWraV0T9VuG
WCLKTVXkcGdtwlfFRjlBz4pYg1htmf5X6DYO8A4jqv2Il9DjXA6USbW1FzXSLr9O
he8Y4IWS6wY7bCkjCWDcRQJMEhg76fsO3txE+FiYruq9RUWhiF1myv4Q6W+CyBFC
Dfvp7OOGAN6dEOM4+qR9sdjoSYKEBpsr6GtPAQw4dy753ec5
-----END CERTIFICATE-----"""

def main(use_stream=True):
    s = _socket.socket()
    ai = _socket.getaddrinfo('micropython.org', 443)
    print('Address infos:', ai)
    addr = ai[0][-1]
    print('Connect address:', addr)
    s.connect(addr)
    s = ssl.wrap_socket(s, cert_reqs=ssl.CERT_REQUIRED, ca_certs=ca_cert_chain)
    print(s)
    s.close()
    return True
main()
```

```
>>> import http_ssl_auth_test
Address infos: [(2, 1, 0, 'micropython.org', ('176.58.119.26', 443))]
Connect address: ('176.58.119.26', 443)
<_SSLSocket 3ffc6540>
```

 I guess both tests work now. Although not sure how to integrate these tests in `multitest`  suite or where exactly put them.

Also for `http_ssl_auth.py`  I guess the cert-chain should be updated before expiration date? 🤔 

## TLDR

So my doubts are **which test to add** to the test suite:
 
Self-signed key, cert tests:
- `ssl_ecc_auth.py` 
- `ssl_rsa_auth.py`
(This needs two esp32  to run as server/client)

Verify example web server (micropython.org):
- `http_ssl_auth.py`
Only one device is needed but cert-chain should (probably) be updated from time to time.

**How I add these** to the test suite?

Carglglz commented:
> Yes I think that would be good. Is that the same default that CPython has?

I think so 
> New in version 3.4.

```
ssl.wrap_socket(sock, keyfile=None, certfile=None, server_side=False, cert_reqs=CERT_NONE, ssl_version=PROTOCOL_TLS, ca_certs=None, do_handshake_on_connect=True, suppress_ragged_eofs=True, ciphers=None)

```

Although this is deprecated since 3.7 by 
``` ssl.create_default_context(purpose=Purpose.SERVER_AUTH, cafile=None, capath=None, cadata=None)```
But in CPython is easier to set security by default since it falls back to system’s default CA certificates. 

For esp-idf there is https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/protocols/esp_crt_bundle.html
that may be worth looking into...

So I guess `MBEDTLS_SSL_VERIFY_NONE` as default is OK for now.

dpgeorge commented:
> **How do I add these** to the test suite?

I think it would be good to add them all.  The multitests in `tests/multi_net/ssl_cert_ecc.py` and `tests/multi_net/ssl_cert_rsa.py`.  And then the other one in `tests/net_inet/ssl_cert.py` (feel free to adjust the test names).

Carglglz commented:
It looks like `tests/multi_net/ssl_cert_ecc.py` and `tests/multi_net/ssl_cert_rsa.py` are failing in some unix ports,  has it anything to do with the fact that unix port uses `axtls` instead of `mbedtls` ? 

Other than that I see no reason why these tests would fail 🤔. The only difference with `tests/multi_net/ssl_data.py` apart from the key/cert is this line:

```diff
 # Server
 def instance0():
@@ -60,7 +58,7 @@ def instance1():
     multitest.next()
     s = socket.socket()
     s.connect(socket.getaddrinfo(IP, PORT)[0][-1])
-    s = ssl.wrap_socket(s)
+    s = ssl.wrap_socket(s, cert_reqs=ssl.CERT_REQUIRED, ca_certs=cert)
     s.write(b"client to server")
     print(s.read(16))
     s.close()
```

Anyway, let me know if I should fix it and how and I will work on it 👍🏼 

And related to the docs let me know if I should explain the new features in more detail.

Apart from that I think this PR is good to go.

dpgeorge commented:
I realised that this PR is not CPython compatible, because in CPython the `ca_certs` argument takes a filename.

I think we should implement `SSLContext` first (see #8915) and then this PR can be added as `SSLContext.load_verify_locations()`.

Carglglz commented:
> I realised that this PR is not CPython compatible, because in CPython the ca_certs argument takes a filename.

True, but there is also some keyword args missing and some have different names 🤔

I agree, we should implement `SSLContext` and make `ussl` CPython compatible, since one of the best features is to be able to write code that runs both in CPython and in MicroPython, which is nice for tests too. 

However, I think it makes more sense to merge this PR first (probably #8916  should go first so `tests/multi_net` can pass properly) as the last "not CPython compatible" and then implement `SSLContext`.

Probably it would be better to implement it in MicroPython and extend functionality in C. 

For reference I've been reviewing [Python TLS/SSL docs](https://docs.python.org/3/library/ssl.html#ssl-contexts), [Python ssl.py](https://github.com/python/cpython/blob/main/Lib/ssl.py) 
and [CPython _ssl.c](https://github.com/python/cpython/blob/main/Modules/_ssl.c#L4021)

and I think this is what could be implemented now in MicroPython:

- [ ] `PROTOCOL_TLS_CLIENT` / `PROTOCOL_TLS_SERVER`
- [ ] `SSLContext.load_cert_chain`
- [ ] `SSLContext.load_verify_locations`
- [ ] `SSLContext.wrap_socket`
- [ ] `SSLContext.protocol`
- [ ] `SSLContext.verify_mode`

Then enhancements in C could be:

- [ ] `SSLContext.check_hostname`

`MBEDTLS_SSL_SERVER_NAME_INDICATION` in `mbedtls_config.h` maybe?

- [ ] Check CERT DATE/TIME (#8854)

(optional)
- [ ] `SSLContext.get_ciphers()` 
- [ ] `SSLContext.set_ciphers(ciphers)`
 
[mbedtls ciphersuites](https://github.com/Mbed-TLS/mbedtls/tree/development/include/mbedtls/ssl_ciphersuites.h)

Finally there is also [`create_default_context`](https://github.com/python/cpython/blob/07aeb7405ea42729b95ecae225f1d96a4aea5121/Lib/ssl.py#L683) and [`_create_unverified_context`](https://github.com/python/cpython/blob/07aeb7405ea42729b95ecae225f1d96a4aea5121/Lib/ssl.py#L721) which is probably the closest to current MicroPython `ssl.wrap_socket`.

dpgeorge commented:
> However, I think it makes more sense to merge this PR first (probably #8916 should go first so `tests/multi_net` can pass properly) as the last "not CPython compatible" and then implement `SSLContext`.

Yes I'd be happy to do that.

> and I think this is what could be implemented now in MicroPython:

Please add that info to #8915 and we can discuss it there.

For this PR: I'd suggest renaming the `ca_certs` keyword arg to something else that doesn't clash with the CPython name, because it takes a different value.  Maybe call it `cadata` to match the argument of `SSLContext.load_verify_locations()`.

Also we need to get the tests working on this PR, after #8916 is merged.

Carglglz committed:
extmod/modussl_mbedtls.c: Switch ca_certs to cadata.

Rename ca_certs kwarg in ssl.wrap_socket to cadata
to avoid name clash with CPython ca_certs.

Signed-off-by: Carlos Gil <carlosgilglez@gmail.com>

Carglglz commented:
> Also we need to get the tests working on this PR, after https://github.com/micropython/micropython/pull/8916 is merged.

Well, now seems all tests are passing  (I had to leave out `ssl_cert_ecc.py` and switch key, cert to `.der` format to match unix mbedtls configuration)

> For this PR: I'd suggest renaming the ca_certs keyword arg to something else that doesn't clash with the CPython name, because it takes a different value. Maybe call it cadata to match the argument of SSLContext.load_verify_locations()

Done in [b90e94c](https://github.com/micropython/micropython/commit/b90e94cde18ebe40f4fd5275abb4be477e3f8953)

> Please add that info to https://github.com/micropython/micropython/issues/8915 and we can discuss it there.

Sure!

> https://github.com/micropython/micropython/pull/8916 (switch unix to mbedtls) is now merged.

Then this PR is ready to be merged, so we can focus on `SSLContext` implementation 👍🏼 .

jimmo commented on tests/net_inet/ssl_cert.py:
```diff
@@ -0,0 +1,121 @@
+    addr = ai[0][-1]
+    print("Connect address:", addr)
+    s.connect(addr)
+    s = ssl.wrap_socket(s, cert_reqs=ssl.CERT_REQUIRED, ca_certs=ca_cert_chain)
```
This test doesn't verify micropython.org. You need to specify server_hostname to do that.

Also, in the set of certs above, you don't need the micropython.org cert. (And we wouldn't want to include micropython.org's cert in a test anyway because it expires every 3 months).

Furthermore, with the default mbedtls behavior, you don't actually need the root cert either. This is a minor CPython inconsistency -- you'd need to add `ssl.VERIFY_X509_PARTIAL_CHAIN` to `SSLContext::verify_flags` to get this behavior, otherwise by default in CPython, you need both the intermediate and the root.

Carglglz replied:
> This test doesn't verify micropython.org. You need to specify server_hostname to do that.

(I just wanted to test cert validation with a web server other than my own self-signed certs.)

True, for that we need to implement `SSLContext.check_hostname`. 
This only validates that the certificate presented by the server matches the certificate chain in `ca_certs`, 

(Also I see now that I forgot to update the kwarg to `cadata`, and I see this is not run by the test suite in CI, so probably I should leave this out too or move it to `examples/network`)

> Also, in the set of certs above, you don't need the micropython.org cert. (And we wouldn't want to include micropython.org's cert in a test anyway because it expires every 3 months).

Yes that was my concern 👀 (see my **TLDR** above)

> Furthermore, with the default mbedtls behavior, you don't actually need the root cert either. This is a minor CPython inconsistency -- you'd need to add ssl.VERIFY_X509_PARTIAL_CHAIN to SSLContext::verify_flags to get this behavior, otherwise by default in CPython, you need both the intermediate and the root.

Then probably `ssl_cert.py` fits better as an example after `SSLContext` implementation. 👍🏼

jimmo replied:
> True, for that we need to implement SSLContext.check_hostname.

wrap_socket already supports the `server_hostname` kwarg (although it isn't documented...).

dpgeorge replied:
It would be good to put this test back.  It's not run as part of CI but it is useful to run on bare-metal to make sure things work (and also of course run on the unix port, so it needs to pass on that as well).

dpgeorge commented on tests/multi_net/ssl_data.py:
```diff
@@ -11,33 +11,10 @@
-)
+cert = b"0\x82\x05\xd70\x82\x03\xbf\xa0\x03\x02\x01\x02\x02\t\x00\xbcc\xb4\x8ap\x0c=I0\r\x06\t*\x86H\x86\xf7\r\x01\x01\x0b\x05\x000\x81\x811\x0b0\t\x06\x03U\x04\x06\x13\x02AU1\x0c0\n\x06\x03U\x04\x08\x0c\x03Foo1\x0c0\n\x06\x03U\x04\x07\x0c\x03Bar1\x140\x12\x06\x03U\x04\n\x0c\x0bMicroPython1\x0c0\n\x06\x03U\x04\x0b\x0c\x03foo1\x160\x14\x06\x03U\x04\x03\x0c\resphome.local1\x1a0\x18\x06\t*\x86H\x86\ ...
+
+key = b"0\x82\t)\x02\x01\x00\x02\x82\x02\x01\x00\xce<\x0fs\n\xb3D2\xce`Z\xb4MJ\xc0\xaa\xfd\x8abC\x13>\xab\r\xcc\x9dDJ\xb7\xd9\xfff\xa6\x81Z\x10\x1d-<\xbdr\x14\n\xfc4\xf8\xc3\xca\xed\xce\x16\xe9R\x83P\xf3\xe0\xe5cC\xf2HP}\x82\xe4\x1bQ\xab\xb5\x15\xcb\xf6\x0eZa\x9f-\xbc\xa8hM\x17L;\tQ\xe2\xc7\xbaWl\x7f\xb0dS\xa3YwU\x81\njLE\xeb\t%\xc8U\xabSw\x85\xdfF\xbf)\x14Xq3\x0f\xf0d\x1a\x10\x1a$\xf0\x83\x0c \x ...
```
Why did you need to change these?

Carglglz replied:
Test suite fails otherwise
> I had to leave out ssl_cert_ecc.py and switch key, cert to .der format to match unix mbedtls configuration

unix port needs changes to allow EC keys/certs and PEM format parsing see my diff comment in #8916
(This is enabled by default in esp-idf)

dpgeorge replied:
These are in DER format already (right?).  And this test does currently pass on unix-to-unix, so I don't understand why it needs to change.

> Test suite fails otherwise

Can you explain exactly what fails?

Carglglz replied:
Oh sorry I misunderstood, I was having problems with `ssl_data.py` old RSA key/cert on my local test suite, mbedtls raising `RSA_BAD_FORMAT` or something like that (I don't remember exactly the message) and decided to change it to see if it worked. But now testing again original RSA keys works. Not sure what exactly happened.
I'm sorry 😞 , reverting changes now...

dpgeorge replied:
Not a problem.  I was also seeing RSA errors on unix with mbedtls and needed to disable `MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED` to get it working.  So that's why it's now working for you with the original RSA keys.

We can get to the bottom of that at some later date.

Carglglz replied:
> I was also seeing RSA errors on unix with mbedtls and needed to disable MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED to get it working

I guess that was it

>  We can get to the bottom of that at some later date.

Yes, I'm curious too 👀

dpgeorge commented on tests/multi_net/ssl_cert_rsa.py:
```diff
@@ -0,0 +1,45 @@
+
+cert = b"0\x82\x05\xd70\x82\x03\xbf\xa0\x03\x02\x01\x02\x02\t\x00\xbcc\xb4\x8ap\x0c=I0\r\x06\t*\x86H\x86\xf7\r\x01\x01\x0b\x05\x000\x81\x811\x0b0\t\x06\x03U\x04\x06\x13\x02AU1\x0c0\n\x06\x03U\x04\x08\x0c\x03Foo1\x0c0\n\x06\x03U\x04\x07\x0c\x03Bar1\x140\x12\x06\x03U\x04\n\x0c\x0bMicroPython1\x0c0\n\x06\x03U\x04\x0b\x0c\x03foo1\x160\x14\x06\x03U\x04\x03\x0c\resphome.local1\x1a0\x18\x06\t*\x86H\x86\ ...
+
+key = b"0\x82\t)\x02\x01\x00\x02\x82\x02\x01\x00\xce<\x0fs\n\xb3D2\xce`Z\xb4MJ\xc0\xaa\xfd\x8abC\x13>\xab\r\xcc\x9dDJ\xb7\xd9\xfff\xa6\x81Z\x10\x1d-<\xbdr\x14\n\xfc4\xf8\xc3\xca\xed\xce\x16\xe9R\x83P\xf3\xe0\xe5cC\xf2HP}\x82\xe4\x1bQ\xab\xb5\x15\xcb\xf6\x0eZa\x9f-\xbc\xa8hM\x17L;\tQ\xe2\xc7\xbaWl\x7f\xb0dS\xa3YwU\x81\njLE\xeb\t%\xc8U\xabSw\x85\xdfF\xbf)\x14Xq3\x0f\xf0d\x1a\x10\x1a$\xf0\x83\x0c \x ...
```
Please format these over multiple lines, using unhexlify like it was done in `ssl_data.py`.

Carglglz replied:
Sure, I will add `server_hostname` to docs too.

dpgeorge commented on docs/library/ssl.rst:
```diff
@@ -31,6 +31,13 @@ Functions
+     Note that for mbedtls based ports, ``ssl.CERT_NONE`` and ``ssl.CERT_OPTIONAL`` will not
+     validate any certificate, only ``ssl.CERT_REQUIRED`` will.
+
+   - *cadata* is the CA certificate chain that will validate the peer's certificate.
```
This takes a concatenation of DER-format certificates, right?  If so would be good to document that here.

Carglglz replied:
Done!

dpgeorge commented on extmod/modussl_mbedtls.c:
```diff
@@ -237,6 +242,20 @@ STATIC mp_obj_ssl_socket_t *socket_new(mp_obj_t sock, struct ssl_args *args) {
+        size_t cacert_len;
+        const byte *cacert = (const byte *)mp_obj_str_get_data(args->cadata.u_obj, &cacert_len);
+        // len should include terminating null
+        ret = mbedtls_x509_crt_parse(&o->cacert, cacert, cacert_len + 1);
```
Currently only DER format is supported.  And in such a case I think this function only parses the first DER in the buffer and will ignore anything after that.

To support a chain of DERs we would need to have a custom DER parser, or at least one that could tell the length of a DER so it could call `mbedtls_x509_crt_parse_der()` multiple times, once for each DER in the chain.

Carglglz replied:
True, I've just checked out `mbedtls_x509_crt_parse()` documentation:

> Parse one DER-encoded or one or more concatenated PEM-encoded certificates and add them to the chained list.
.... 
The buffer holding the certificate data in PEM or DER format. For certificates in PEM encoding, this may be a concatenation of multiple certificates; for DER encoding, the buffer must comprise exactly one certificate.

And `mbedtls_x509_crt_parse_der()`:
> Parse a single DER formatted certificate and add it to the end of the provided chained list.

Which seems like a concatenated DER-encoded certificates file does not exist. 

> custom DER parser, or at least one that could tell the length of a DER so it could call mbedtls_x509_crt_parse_der() multiple times, once for each DER in the chain

I guess you could parse multiple files/buffers calling `mbedtls_x509_crt_parse_der() ` multiple times, but from a single file I don't think is possible, you would need to add a header with the sizes or some delimiters like PEM. 

Probably is better to support PEM-encoding or make it multiple files/list of buffers than a single "chain DER file".
In any case I think this fits better in `SSLContext.load_verify_locations`, right?

dpgeorge replied:
Let's solve this problem later on, as part of SSLContext.  For now what is done in this PR is a good start, and a subset of what we could eventually support (either PEM or list/tuple/concat'd DER).

dpgeorge commented on tests/multi_net/ssl_cert_rsa.py:
```diff
@@ -0,0 +1,143 @@
+# This self-signed key/cert pair is randomly generated and to be used for
+# testing/demonstration only.  You should always generate your own key/cert.
+
+cert = binascii.unhexlify(
```
Can you please include as a comment the command use to generate this cert and key.

dpgeorge commented on tests/net_inet/ssl_cert.py:
```diff
@@ -0,0 +1,65 @@
+    import binascii
+
+
+ca_cert_chain = binascii.unhexlify(
```
Is this just the Let's Encrypt cert?  Can you please include as a comment the command you used to generate this data.

Carglglz replied:
Yes it's this one:
```
1 s:/C=US/O=Let's Encrypt/CN=R3
   i:/C=US/O=Internet Security Research Group/CN=ISRG Root X1
```

dpgeorge commented on tests/net_inet/ssl_cert.py:
```diff
@@ -0,0 +1,65 @@
+    s = ssl.wrap_socket(
+        s, cert_reqs=ssl.CERT_REQUIRED, cadata=ca_cert_chain, server_hostname="micropython.org"
+    )
+    print(s)
```
This won't output a consistent result, it'll have a different address each time.

Instead, I suggest to do this:
```python
s.write(b"GET / HTTP/1.0\r\n\r\n")
print(s.read(17))
```
And then provide a `ssl_cert.py.exp` file that contains the expected output of the test, ie:
```
b'HTTP/1.1 200 OK\r\n'
```
Then the test should pass on the unix port (at least).

Carglglz replied:
You are right, done.

dpgeorge commented:
Rebased, squashed and merged in b41cfea02ab9f566785da5e5e0863d009f3ac527, with some minor tweaks to the docs.

Thank you to all involved!
